### PR TITLE
Cascading Merkle Tree

### PR DESCRIPTION
**Pull request by ewoolsey**
Opening this up as a draft! I think this is pretty close to where I would like it to be. Please let me know if you guys have any suggestions here!

### CODE REVIEW

File: src/cascading_merkle_tree.rs [hunk 1]
```diff
@@ -0,0 +1,1458 @@
...
+/// of the main tree.
+///
+/// This function assumes that storage is already initialized with empty
+/// values and is the correct length for the subtree.
+/// If 'leaves' is not long enough, the remaining leaves will be left empty
+///
+/// storage.len() must be a power of 2 and greater than or equal to 2
+/// storage is 1 indexed
+///
+/// ```markdown
```
Comment (Dzejkop) [inline src/cascading_merkle_tree.rs:906]: Unclosed code block, also `makrdown` is probably not necessary

File: src/cascading_merkle_tree.rs [hunk 2]
```diff
@@ -0,0 +1,1458 @@
...
+    ///
+    /// # Panics
+    ///
+    /// Panics if the leaf index is not less than the current
+    /// number of leaves.
+    pub fn set_leaf(&mut self, leaf: usize, value: H::Hash) {
+        assert!(leaf < self.num_leaves(), "Leaf index out of bounds");
+        let index = index_from_leaf(leaf);
+        self.storage[index] = value;
+        self.storage.propogate_up(index);
```
Comment (Dzejkop) [inline src/cascading_merkle_tree.rs:111]: typo

File: src/cascading_merkle_tree.rs [hunk 3]
```diff
@@ -0,0 +1,1458 @@
...
+        }
+        if len < 2 {
+            bail!("Storage length must be greater than 1");
+        }
+
+        let width = len >> 1;
+        let depth = width.ilog2() as usize;
+
+        let num_leaves = self.num_leaves();
+        let start = index_from_leaf(num_leaves);
```
Comment (Dzejkop) [inline src/cascading_merkle_tree.rs:491]: Maybe `last_leaf` instead of `start`?

File: src/cascading_merkle_tree.rs [hunk 4]
```diff
@@ -0,0 +1,1458 @@
...
+        hash
+    }
+
+    /// Validates all elements of the storage, ensuring that they
+    /// correspond to a valid tree.
+    pub fn validate(&self) -> Result<()> {
+        self.storage.validate(&self.empty_value)
+    }
+
+    // pub fn extend_from_slice(&mut self, leaves: &[H::Hash]) {
```
Comment (Dzejkop) [inline src/cascading_merkle_tree.rs:281]: Should be deleted?

File: src/cascading_merkle_tree.rs (hunk 1 above)
Comment (ewoolsey) [reply src/cascading_merkle_tree.rs:906]: So I struggled with this haha. If you don't put a code block, rust get's confused by the format and thinks it's a doc test and explodes. Same thing happens if you put a code block without a tag. But if you put a code block with a tag it's fine. If you have a better solution let me know haha.

File: src/cascading_merkle_tree.rs (hunk 4 above)
Comment (ewoolsey) [reply src/cascading_merkle_tree.rs:281]: I'm going to implement this in the next few days if it's useful for the world tree service. Will either delete or implement before we merge here!

File: src/cascading_merkle_tree.rs (hunk 2 above)
Comment (ewoolsey) [reply src/cascading_merkle_tree.rs]: fixed!

File: src/cascading_merkle_tree.rs (hunk 3 above)
Comment (ewoolsey) [reply src/cascading_merkle_tree.rs]: It's technically not the last leaf, it's the first_empty haha. I'll rename it to that.

File: src/cascading_merkle_tree.rs (hunk 1 above)
Comment (Dzejkop) [reply src/cascading_merkle_tree.rs:908]: Oh I see, yeah that makes sense

File: src/cascading_merkle_tree.rs [hunk 5]
```diff
@@ -0,0 +1,1464 @@
...
+///
+/// Leaves are 0 indexed
+/// 0  1  2  3  4  5  6  7
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct CascadingMerkleTree<H: Hasher, S: CascadingTreeStorage<H> = Vec<<H as Hasher>::Hash>> {
+    depth:         usize,
+    root:          H::Hash,
+    empty_value:   H::Hash,
+    sparse_column: Vec<H::Hash>,
```
Comment (Dzejkop) [inline src/cascading_merkle_tree.rs:41]: Could we add a little comment here explaining what this field is? As I understand these are the values of empty subtrees by depth

**Review by Dzejkop — APPROVED** [review]

**Review by 0xKitsune — APPROVED** [review]

**Pull request merged**